File: Python/1055. Shortest Way to Form String.py
class Solution:
    def shortestWay(self, source: str, target: str) -> int:
        src = [[] for i in range(26)]
        for (i, c) in enumerate(source):
            src[ord(c) - ord('a')].append(i)            
        count = 1
        start = 0
        i = 0
        for i in range(len(target)):
            idx = src[ord(target[i])-ord('a')]
            if len(idx) == 0:
                return -1
            curr = bisect.bisect_left(idx, start)             
            if curr == len(idx):
                count += 1
                start = idx[0] + 1
            else:
                start = idx[curr] + 1              
        return count
                

File: 2021/1055. Shortest Way to Form String.cpp
class Solution {
public:
    int shortestWay(string source, string target) {
        int index = 0;
        for (int i = 0; i < target.length(); i++) {
            int start = index;
            while (index < start + source.length() && target[i] != source[index % source.length()]) {
                index++;
            }            
            if (index >= start + source.length()) {
                return -1;
            }
            index++;
        }
        return (index + source.length() - 1) / source.length();
    }
};